### PR TITLE
tests: e2e validation of plugin field updates for compatibility

### PR DESCRIPTION
**Pull request by omegabytes**
Recreates end-to-end tests used to validate updates to plugin fields that can occur when the version of a control plane and a data plane differ. These tests were removed in #368 due to a refactor.

### CODE REVIEW

File: internal/test/e2e/version_compatibility_test.go
Comment (GGabriele) [inline internal/test/e2e/version_compatibility_test.go:195]: I think this can be rewritten in a more efficient way:

```
    for _, pluginTest := range VersionCompatibilityOSSPluginConfigurationTests {
		if pluginTest.FieldUpdateChecks == nil {
			continue
		}
		if len(pluginTest.VersionRange) > 0 {
			version := versioning.MustNewRange(pluginTest.VersionRange)
			if !version(dataPlaneVersion) {
				continue
			}
		}
		expectedPluginsMap[pluginTest.Name] = pluginTest
	}
```

This way you don't need the extra `addPlugin` and avoid cycles when not needed

File: internal/test/e2e/version_compatibility_test.go
Comment (GGabriele) [inline internal/test/e2e/version_compatibility_test.go:212]:
```suggestion
	// create plugins
```

File: internal/test/e2e/version_compatibility_test.go
Comment (GGabriele) [inline internal/test/e2e/version_compatibility_test.go:216]:
```suggestion
			require.NoError(t, json.ProtoJSONUnmarshal([]byte(pluginTest.Config), &config))
```

File: internal/test/e2e/version_compatibility_test.go
Comment (GGabriele) [inline internal/test/e2e/version_compatibility_test.go:228]:
```suggestion
		require.NoError(t, err)
```

File: internal/test/e2e/version_compatibility_test.go
Comment (GGabriele) [inline internal/test/e2e/version_compatibility_test.go:233]:
```suggestion
	// fetch list of plugins from the DP
```

File: internal/test/e2e/version_compatibility_test.go
Comment (GGabriele) [inline internal/test/e2e/version_compatibility_test.go:240]: You can replace this with the other check you have at line 250:

```
require.Equal(t, len(expectedPluginsMap), len(dataPlanePlugins), "plugins configured count does not match")
```

File: internal/test/e2e/version_compatibility_test.go
Comment (GGabriele) [inline internal/test/e2e/version_compatibility_test.go:250]: This can be removed if you move it up in the code, as per my other comment

File: internal/test/e2e/version_compatibility_test.go
Comment (GGabriele) [inline internal/test/e2e/version_compatibility_test.go:254]:
```suggestion
			require.Contains(t, dpPluginsMap, name, "plugin not present in the DP")
```

File: internal/test/e2e/version_compatibility_test.go
Comment (GGabriele) [inline internal/test/e2e/version_compatibility_test.go:259]: I think these can simply be (please double-check it!):

```
						want := update.Value
						have := dpPluginsMap[name].Config[update.Field]
```

Comment (omegabytes) [reply internal/test/e2e/version_compatibility_test.go:259]: I fumbled around with this for a bit before landing on the presented solution. Here's what happens if I don't stringify and compare:
```
Error:  Not equal: 
            expected: []string([]string{"kong.log.err('Hello Koko!')"})
            actual  : []interface {}([]interface {}{"kong.log.err('Hello Koko!')"})
```

Comment (omegabytes) [reply internal/test/e2e/version_compatibility_test.go:259]: I got some feedback from @tjasko as well. Here's the two paths I can take:
```
for _, update := range updates {
	want := update.Value
	have := dpPluginsMap[name].Config[update.Field]
	switch update.Value.(type) {
	case string:
		require.Equal(t, want, have, "failed to validate plugin updates")
	case []string:
		require.ElementsMatch(t, want, have, "failed to validate plugin updates")
	default:
		require.Fail(t, fmt.Sprintf("Unexpected type(s): want: %s, have: %s, ",
		    reflect.TypeOf(want), reflect.TypeOf(have)))
	}
}
```
Or the much less verbose:
```
for _, update := range updates {
	want := fmt.Sprintf("%v", update.Value)
	have := fmt.Sprintf("%v", dpPluginsMap[name].Config[update.Field])
	require.Equal(t, want, have, "failed to validate plugin updates")
}	
```
I'd rather go with the stringified version since these version compatibility handlers have so many conditional statements already. I'm making sure that I'm not introducing something flaky by taking a shortcut.

Comment (GGabriele) [reply internal/test/e2e/version_compatibility_test.go:256]: I believe we can happily accept verbosity for correctness. If we are absolutely sure that stringify-all-the-things works well in all cases, including the `[]string` one, then proceed. Otherwise, let's just write some more lines of code. You can always wrap the verification logic in an util function to avoid more nesting:

```
for _, update := range updates {
	want := update.Value
	have := dpPluginsMap[name].Config[update.Field]
	verifyUpdates(want, have) // or something like `require.NoError(t, verifyUpdates(want, have))`
}
```

Comment (omegabytes) [reply internal/test/e2e/version_compatibility_test.go]: >  can't we simply convert the values (expected and response) to JSON and compare that way?

I considered this, but this is also converting to a string so I would (lazily) just skip the marshaling steps.

I pushed an update that checks the type instead.

File: internal/test/e2e/version_compatibility_test.go
Comment (omegabytes) [inline internal/test/e2e/version_compatibility_test.go:251]: I added this test runner to more verbosely describe which plugin(s) fail and why. This includes the expected and actual config diffs. Previously, we were only returning a list of plugin names that failed.

File: internal/test/e2e/version_compatibility_test.go
Comment (omegabytes) [inline internal/test/e2e/version_compatibility_test.go:258]: I just noticed this wont work when the Field value uses a path.to.field format. Eg: `update.Field = "response.body.time"`. I'll have to rewrite this using gjson or something...

File: internal/test/e2e/version_compatibility_test.go
Comment (mikefero) [reply internal/test/e2e/version_compatibility_test.go]: Your call, but this is going to be a maintenance issue in the future as we will need to keep adding to the case statement for each new type added.

Leaving final decision for @GGabriele.

Comment (GGabriele) [reply internal/test/e2e/version_compatibility_test.go]: Yeah I agree, I think you can easily re-use part of [the old logic](https://github.com/Kong/koko/pull/368/files#diff-5351ee7e6ddda0e9a38eee6852f369ac1c1fd272bba60c61f6ced419486215c3L456)...something like (definitely needs to be refactored):

```
                         for rawVersion, updates := range expectedPlugin.FieldUpdateChecks {
				parsedVersion := versioning.MustNewRange(rawVersion)
				config, err := json.ProtoJSONMarshal(dpPluginsMap[name])
				require.NoError(t, err, fmt.Sprintf("marshal %s plugin config: %v", name, err))
				configStr := string(config)
				if parsedVersion(dataPlaneVersion) {
					for _, update := range updates {
						res := gjson.Get(configStr, update.Field)
						if !res.Exists() || res.Value() != update.Value {
							return fmt.Errorf("expected %s, but got %s", update.Value, res.Value())
						}
					}
				}
			}
```

**Review by mikefero — APPROVED** [review]
Looks good to me, but I would still prefer not to validate differently based on types. Please get another +1 from @GGabriele.

**Review by GGabriele — APPROVED** [review]

**Review by tjasko — APPROVED** [review]

File: internal/test/e2e/version_compatibility_test.go
Comment (mikefero) [reply internal/test/e2e/version_compatibility_test.go]: You should be able to use [require.JSONEq](https://pkg.go.dev/github.com/stretchr/testify/require#JSONEq) to compare the JSON component instead of the objects themselves.

Comment (omegabytes) [reply internal/test/e2e/version_compatibility_test.go]: Thanks for the feedback. Looks like the solution with require.JSONEq will require creating of test plugin config fixtures to compare against. I'm going to stick with the approved solution, and we can revisit this if adding checks on types becomes unwieldy to maintain 😄

**Pull request merged**
